model/adjuster/clockskew.go

```diff
@@ -0,0 +1,196 @@
+//
+// The algorithm assumes that all spans have unique IDs, otherwise it returns
+// an error, so the trace may need to go through another adjuster first.
+func ClockSkew() Adjuster {
+	return Func(adjustClockSkew)
+}
+
+func adjustClockSkew(trace *model.Trace) (*model.Trace, error) {
+	adjuster := &clockSkewAdjuster{
+		trace: trace,
+	}
+	adjuster.mapIDsToSpans()
+	adjuster.buildSubGraphs()
+	for _, n := range adjuster.roots {
+		skew := clockSkew{hostKey: n.hostKey}
+		adjuster.adjustNode(n, nil, skew)
+	}
+	return adjuster.trace, multierror.Wrap(adjuster.errors)
+}
+
+type node struct {
+	span     *model.Span
+	children []*node
+	hostKey  string
+}
+
+// hostKey returns a string representation of the host identity that can be used
+// to determine if two spans originated from the same host.
+//
+// TODO convert process tags to a canonical format somewhere else
+func hostKey(span *model.Span) string {
+	if tag, ok := span.Process.Tags.FindByKey("ip"); ok {
+		if tag.VType == model.StringType {
+			return tag.VStr
+		}
+		if tag.VType == model.Int64Type {
+			var buf [4]byte
+			ip := buf[0:4]
+			binary.BigEndian.PutUint32(ip, uint32(tag.Int64()))
+			return net.IP(ip).String()
+		}
+		if tag.VType == model.BinaryType {
+			if l := len(tag.Binary()); l == 4 {
+				return net.IP(tag.Binary()).String()
+			} else if l == 16 {
+				ip := net.IP(tag.Binary())
+				if ipv4 := ip.To4(); ipv4 != nil {
+					return ipv4.String()
+				}
+				return ip.String()
+			}
+		}
+	}
+	return ""
+}
+
+type clockSkewAdjuster struct {
+	trace  *model.Trace
+	spans  map[model.SpanID]*node
+	roots  map[model.SpanID]*node
+	errors []error
+}
+
+type clockSkew struct {
+	delta   int64
+	hostKey string
+}
+
+// mapIDsToSpans builds a map of span IDs -> node{}.
+func (a *clockSkewAdjuster) mapIDsToSpans() {
+	a.spans = make(map[model.SpanID]*node)
+	for _, span := range a.trace.Spans {
+		if _, ok := a.spans[span.SpanID]; ok {
+			a.errors = append(a.errors, fmt.Errorf("found more than one span with ID=%x", span.SpanID))
+		} else {
+			a.spans[span.SpanID] = &node{
+				span:    span,
+				hostKey: hostKey(span),
+			}
+		}
+	}
+}
+
+// finds all spans that have no parent, i.e. where parentID is either 0
+// or points to an ID for which there is no span.
+func (a *clockSkewAdjuster) buildSubGraphs() {
+	a.roots = make(map[model.SpanID]*node)
+	for _, n := range a.spans {
+		// TODO handle FOLLOWS_FROM references
+		if n.span.ParentSpanID == 0 {
+			a.roots[n.span.SpanID] = n
+			continue
+		}
+		if p, ok := a.spans[n.span.ParentSpanID]; ok {
+			p.children = append(p.children, n)
+		} else {
+			err := fmt.Errorf("invalid parent span ID=%x in the span with ID=%x", n.span.ParentSpanID, n.span.SpanID)
+			a.errors = append(a.errors, err)
+			// Treat spans with invalid parent ID as root spans
+			a.roots[n.span.SpanID] = n
+		}
+	}
+}
+
+func (a *clockSkewAdjuster) adjustNode(n *node, parent *node, skew clockSkew) {
+	if n.hostKey != skew.hostKey || n.hostKey == "" {
```

Comment:
You can run into a nil ptr exception here if you're not careful. If n.hostKey == "" and the parent node is nil (which means skew.hostKey == ""), calculateSkew function will panic since parent is nil.